### PR TITLE
Add @bugsnag/delivery-expo module

### PR DESCRIPTION
**Pull request by bengourley**
This PR adds an Expo specific delivery module which is the default delivery mechanism for the new `@bugsnag/expo` notifier.

#### Added a new module `@bugsnag/delivery-expo`

This module implements the `sendReport()` and `sendSession()` functions as required by this interface. Additionally it starts polling for unsent payloads when it is initialised.

`@bugsnag/delivery-expo` was added as a dependency of `@bugsnag/expo`.

#### Refactored the delivery interface

Previously the delivery mechanisms had no context or state when they were first initialised. They only received things like the `logger` and `config` as arguments to `sendReport|Session()`. Since the expo delivery mechanism potentially needs to start sending reports _before_ any reports/sessions are sent, the `client` now passes itself in when the delivery mechanism is initialised.

The actual impact of this change is pretty small, but because many of the unit tests use the delivery mechanism to make assertions, it appears large because many of these had to change as a result.

### Discussion

The high-level overview of `packages/expo-delivery/delivery.js`'s logic is as follows:

- attempt to send a report/session
- if unsuccessful, stop here!
- do we have a status code that means we shouldn't try again? e.g anything in the range `40x`
- if yes, stop here!
- enqueue the failed payload

The high level overview of `packages/expo-delivery/redelivery.js`'s logic is as follows:

- begin loop
  - attempt to dequeue a payload that failed to send
  - if result of dequeue is null, wait `n` ms and then goto "begin loop"
  - attempt to resend the failed payload
  - if successful, goto "begin loop" immediately
  - if unsuccessful, and the reason can be retried, re-enqueue the item, wait for `n` ms then goto "begin loop"
  - if unsuccessful, and the reason cannot be retried or the item has reached its `m` attempts then do not re-enqueue, wait `n` ms and then goto "begin loop"

I chose the values `n=30000` and `m=5`, how do people feel about those?

I considered using the [`NetInfo` api](https://docs.expo.io/versions/v28.0.0/react-native/netinfo/) to only attempt sending when the device appears online, but this requires an app permission on android that our users might not have enabled.

The high level overview of `packages/expo-delivery/queue.js`'s logic is as follows:

- within the app's cache directory, use the directory structure `bugsnag/payloads/…`.
- in that directory, store json files that represent failed requests using the naming pattern `bugsnag-payload-${isodate}-${uid}.json` so that when it is lexicograhically sorted, the items appear in date order. For the purpose of this implementation, failed reports created in exactly the same millisecond are considered equal and do not matter which order they are sorted in. The purpose of the `bugsnag-payload-` part which at first glance may seem like information that is already encoded in the path is in case any other files are accidentally or automatically written to this directory. It means we can tolerate their existence.
- to enqueue a payload, ensure the directory structure exists, json stringify the payload and write it using the naming pattern
- to dequeue a payload, read the directory, sort the list lexicographically, take the first name and json parse the contents then delete the file. If no files exists, return null.

### Concerns

What happens if there is a bug in this persistence logic? What happens if the file system gets into a corrupt state? How do other notifiers deal with these problems?

### Testing

I've written unit tests that cover as much of this new code as is reasonably possible. A summary of the coverage report is as follows:

image

To test out this functionality on a device/emulator, follow the steps in `test/expo/fixtures/test-app`. Switch the device into aeroplane mode and trigger some errors, then disable aeroplane mode. You should see errors reported to the dashboard after up to 30s delay.

### CODE REVIEW

**Comment by bengourley** [comment]
Thanks for the review @fractalwrench. Really helpful comments. I'll respond to each one below to shake out what to act on.

> On Android/Cocoa for fatal errors we write reports to disk immediately, then attempt delivery. Are there any scenarios where we might need to do this in Expo, rather than attempting to make a request first?

I'm not 100% sure on this one. We should know a bit better when there are different end-to-end tests for it, but I'll make a note to chat to @Cawllec about it so it's definitely considered.

> Is there any way of adding the permission onto the AndroidManifest automatically?

I think we could do this automatically but I don't think we should. The developer might not want this for a good reason. Equally, it adds some complexity without a huge benefit. The device might report itself online when there isn't actually a network connection. The current implementation is a little crude but very simple, and I think it would be effective. It doesn't make any assumptions about the network, but when a report gets through, it tries another immediately after.

> We should add some limit to the number of stored reports and delete the oldest one, as per Android/Cocoa

Agreed. What should the limit be? Are Android/Cocoa the same?

> The status code check seems fairly broad and would discard some requests that could be retried (e.g. 429)

Great spot. I'll review the list of status codes that shouldn't be retried on and make this check more appropriate.

> If the app is being shutdown and an error is scheduled by redelivery.js, will the error be persisted on disk if redelivery fails?

I don't have a good answer for this. It depends how forcefully the app is shutdown. If the event loop is allowed to clear, then it will get persisted, but if it is terminated forcefully it could be lost. Again, a good one to end-to-end test if we can?

> We may want to separate session/error payloads into separate directories, as this would allow us to limit/prioritise delivery independently

Sounds reasonable but adds complexity – do you think it's worth it?

> If a file isn't valid JSON, queue.js should delete it

Agreed.

> I ran the example app on an iOS + Android simulator/emulator with a network connection and got the following message logged every X seconds [omited]

Ahh, I see why this is. The directory is lazily created (the first time a payload is enqueued). We could supress the log, or ensure the directory exists at startup. What do you think?

To summarise, these are what I think the following code changes should be:

- Add a hard limit to how many reports are stored on disk. Value TBD.
- Update the logic around non-retryable status codes to allow 429 (and others) to be retried.
- If `JSON.parse()` fails, delete the file
- Update logic to avoid `[bugsnag], Error redelivering payload` message when the directory does not exist. Supress log or create directory ahead of time, TBD.

**Comment by fractalwrench** [comment]
> > On Android/Cocoa for fatal errors we write reports to disk immediately, then attempt delivery. Are there any scenarios where we might need to do this in Expo, rather than attempting to make a request first?
> 
> I'm not 100% sure on this one. We should know a bit better when there are different end-to-end tests for it, but I'll make a note to chat to @Cawllec about it so it's definitely considered.

I think we need to answer this before doing any further development, as it could result in substantial changes to the implementation. If an unhandled error/promise rejection kills the app process, then we would need to write to disk immediately as otherwise we might encounter data loss.

> > Is there any way of adding the permission onto the AndroidManifest automatically?
> 
> I think we could do this automatically but I don't think we should. The developer might not want this for a good reason. Equally, it adds some complexity without a huge benefit. The device might report itself online when there isn't actually a network connection. The current implementation is a little crude but very simple, and I think it would be effective. It doesn't make any assumptions about the network, but when a report gets through, it tries another immediately after.

We check the network connection on Android & iOS because it avoiding unnecessary use of the device radio, which prolongs battery life. Unless there's a technical reason preventing us from adding the permission to the AndroidManifest then I think we need to perform this check, and also flush reports when a connection is regained.

> > We should add some limit to the number of stored reports and delete the oldest one, as per Android/Cocoa
> 
> Agreed. What should the limit be? Are Android/Cocoa the same?

The max limit is 128 for Android and 5 for Cocoa so a bit inconsistent. I would say go with 128 and we can always change that limit later on if needed.

> > If the app is being shutdown and an error is scheduled by redelivery.js, will the error be persisted on disk if redelivery fails?
> 
> I don't have a good answer for this. It depends how forcefully the app is shutdown. If the event loop is allowed to clear, then it will get persisted, but if it is terminated forcefully it could be lost. Again, a good one to end-to-end test if we can?

I think this is another thing we need to answer before any further development takes place. Android/Cocoa delete the file after it has been delivered, to prevent this sort of data loss.

> > We may want to separate session/error payloads into separate directories, as this would allow us to limit/prioritise delivery independently
> 
> Sounds reasonable but adds complexity – do you think it's worth it?

Yes. We might want to migrate error payloads but not session payloads (which is admittedly easier with dynamic typing), prioritise the delivery of errors over sessions, etc. This approach would also be consistent with Android/Cocoa.

> To summarise, these are what I think the following code changes should be:
> 
> * Add a hard limit to how many reports are stored on disk. Value TBD.
> * Update the logic around non-retryable status codes to allow 429 (and others) to be retried.
> * If `JSON.parse()` fails, delete the file
> * Update logic to avoid `[bugsnag], Error redelivering payload` message when the directory does not exist. Supress log or create directory ahead of time, TBD.

These changes sound good 👍

**Comment by bengourley** [comment]
Ok, after much poking around, here are the concrete changes I'm going to make…

### Write to disk when app would reload due to an error

In spite of the fact that Expo allows any async execution in the global error handler to complete before it reloads, we will do the following: __If an error is being reported that will cause the app to reload, immediately save it to disk rather than attempt to send it.__

This means that…
1. a slow network request won't stop the crashy app from reloading in a timely manner
2. it prevents the user from being able to force quit the app before the request has failed and the payload saved to disk
3. it is consistent with other mobile notifiers

Manual calls to `notify()` and unhandled errors which wouldn't cause the app to reload should be attempted.

### Undelivered payloads should exist on disk until successfully delivered 

The current implementation optimistically removes an item from the queue and only re-adds it if the subsequent attempt fails. This logic should be reversed so that the item remains on disk until it is successfully sent. This prevents the possibility of a forcibly-closed app losing an undelivered payload, but it introduces the risk of delivering a payload multiple times (if the app was forcibly-closed after the payload was delivered but before it was deleted).

### Separate storage of reports/sessions

Use different directories for reports/sessions and different redelivery loops.

### Limit the number of undelivered items in a directory

Each directory should be allowed no more than `128 / 2 = 64` undelivered items. A new payload should replace the oldest payload if this limit is reached.

### Delete a saved payload that fails `JSON.parse(payload)`

If it's not JSON, it's useless. It's taking up space, so delete it.

### Enumerate the list of HTTP status codes that are not retryable

Currently `429` would not be retried. There are probably others.

### Assume `NetInfo` API exists

I did some more digging on this topic and it's good news – `android.permission.ACCESS_NETWORK_STATE` is enabled by default on all Expo apps, and it also exists in the minimal list of permissions you can set it in Expo app, so we can just go ahead and use it.

This means we can update the implementation to respond to `connectionChange` events, rather than an arbitrary timeout.

**Comment by fractalwrench** [comment]
I'm happy with the proposed approach to the changes.

File: packages/delivery-expo/queue.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    await FileSystem.makeDirectoryAsync(this._path, { intermediates: true })
+  }
+
+  /*
+   * Keeps the queue size bounded by MAX_LENGTH
+   */
+  async _truncate () {
+    // this isn't atomic so only enter this method once at any time
+    if (this._truncating) return
+    this._truncating = true
```
[the remark this answers is not in the document]
Comment (bengourley) [reply packages/delivery-expo/queue.js]: JS is single threaded 😁

File: packages/delivery-expo/delivery.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      const url = client.config.endpoints.sessions
+
+      let body, opts
+      try {
+        body = payload.session(session, client.config.filters)
+        opts = {
+          method: 'POST',
+          headers: {
+            'Content-Type': 'application/json',
+            'Bugsnag-Api-Key': client.config.apiKey,
```
[the remark this answers is not in the document]
Comment (bengourley) [reply packages/delivery-expo/delivery.js]: I did consider this and I think it's fine. All errors before the change would go the old project and all errors after the change would go to the new project. That seems reasonable to me.

Plus if you made a change – say moving from on prem to bugsnag.com, removing some PII from from the payload in the process – it would be surprising if some old errors with all the PII went to bugsnag.com.

File: packages/delivery-expo/delivery.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+          method: 'POST',
+          headers: {
+            'Content-Type': 'application/json',
+            'Bugsnag-Api-Key': report.apiKey || client.config.apiKey,
+            'Bugsnag-Payload-Version': '4',
+            'Bugsnag-Sent-At': isoDate()
+          },
+          body
+        }
+        if (!networkStatus.isConnected || (report._handledState && report._handledState.attemptDelivery === false)) {
```
[the remark this answers is not in the document]
Comment (bengourley) [reply packages/delivery-expo/delivery.js]: Yeah I pondered over where the best place to attach this information to the report was, and of all the existing places I think it's here. But I agree it's not perfect.

Basically the `react-native-global-error-handler` needs to communicate _"I caught this error which would have caused the app to reload"_. `unhandled: true` is the case for stuff like render errors caught in an react error boundary, which wouldn't cause that, so we can't just base this off of `unhandled` unfortunately – we need the extra info.

File: packages/plugin-react-native-global-error-handler/error-handler.js [hunk 4]
```diff
@@ -17,7 +17,8 @@ module.exports = {
       client.notify(createReportFromErr(error, {
         severity: 'error',
         unhandled: true,
-        severityReason: { type: 'unhandledException' }
+        severityReason: { type: 'unhandledException' },
+        attemptDelivery: false
```
[the remark this answers is not in the document]
Comment (bengourley) [reply packages/plugin-react-native-global-error-handler/error-handler.js]: explained above, hopefully it makes sense.

File: packages/delivery-expo/queue.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (fractalwrench) [reply packages/delivery-expo/queue.js]: Maybe we can talk this through tomorrow - I'm not entirely sure why there's a need for a flag if concurrent access to the method isn't expected. Putting it in the terms of threads was probably a bit misleading on my part

File: packages/delivery-expo/delivery.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (fractalwrench) [reply packages/delivery-expo/delivery.js]: The scenario I'm thinking of would be:

1. error captured and stored on disk
2. API key changes before next launch
3. stored error is sent using the new API key, rather than the old one

File: packages/delivery-expo/queue.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (snmaynard) [reply packages/delivery-expo/queue.js]: Concurrent access is expected, but js only switches execution contexts on i/o and so it will only execute another context later in this function when the promises are used to read the filesystem. So you can maybe think of this part as atomically executed

File: packages/delivery-expo/delivery.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (fractalwrench) [reply packages/delivery-expo/delivery.js]: My gut feeling is that it would make more sense as `report.attemptDelivery`, as whether delivery should be attempted is a property of a report, rather than its handledState.

Is there any risk that including this value in the `report` could result in it being serialised?

File: packages/plugin-react-native-global-error-handler/error-handler.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (fractalwrench) [reply packages/plugin-react-native-global-error-handler/error-handler.js]: resolving this one as I wrote the same thing twice

File: packages/delivery-expo/queue.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (fractalwrench) [reply packages/delivery-expo/queue.js]: That makes sense, I don't think there's anything to address here in that case.

File: packages/delivery-expo/delivery.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (bengourley) [reply packages/delivery-expo/delivery.js]: That wouldn't be the case – this block of code is only ever entered once for an error report. When it is taken from disk it goes via the lower level `send()` function in this file, not this `sendReport()` function.

File: packages/delivery-expo/queue.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    this._truncating = false
+  }
+
+  /*
+   * Ensure the persistent cache directory exists
+   */
+  async init () {
+    const { exists, isDirectory } = await FileSystem.getInfoAsync(this._path)
+    if (exists && isDirectory) return true
+    await FileSystem.makeDirectoryAsync(this._path, { intermediates: true })
```
[the remark this answers is not in the document]
Comment (bengourley) [reply packages/delivery-expo/queue.js]: yeah the return value isn't important here. I'll remove `true` because it's misleading

File: packages/delivery-expo/delivery.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (bengourley) [reply packages/delivery-expo/delivery.js]: > Is there any risk that including this value in the report could result in it being serialised?

Nope, we manually describe the properties that get serialised [here](https://github.com/bugsnag/bugsnag-js/blob/2091b7e62e2e073bcbbd5f90d3ff29a2d813672a/packages/core/report.js#L94-L118).

I think I'd be happy to move to your alternative suggestion so I'll make that change.
